### ISSUE TITLE
[BUG] Not receiving StoreResponse.Error.Exception from exceptions thrown in Fetcher when calling store.get(key)

### ISSUE THREAD

**Issue by TylerMcCraw** [dropbox/Store#280]
**Describe the bug**
When there is no data in the SourceOfTruth and the Fetcher throws an exception, a collector of `store.get(key)` only receives StoreResponse.Loading and does not receive StoreResponse.Error.Exception.

To be specific, my SourceOfTruth is a Realm that listens to a single object and my Fetcher call is a single API call using Retrofit to fetch this specific object.
In my use case, there is no data in my Realm for this object, so the Store knows to use the Fetcher and make a call to fetch the data. The fetcher call fails (I'm expecting this) with a 404 HTTP Not Found exception because the user doesn't have access to the data or because the data doesn't exist on the server.
The issue is that when I make a call to `store.get(key)`, the collector only receives a StoreResponse.Loading emission.
This causes issues because I need to be able to properly handle the error from the Fetcher where my collection happens and the store.get() call actually waits for a non-Loading and non-NoNewData result, which leads to the call being indefinitely suspended.

**Expected behavior**
When there is no data in the SourceOfTruth and the Fetcher throws an exception, a collector of `store.get(key)` should receive StoreResponse.Loading and then StoreResponse.Error.Exception.

Sample code (simplified):
```
    // In repository
    private val userStore: Store<String, DataSourceResult<User>> = StoreBuilder.from(
        fetcher = Fetcher.of { userId ->
            userRemoteDataSource.getUser(userId)
        },
        sourceOfTruth = SourceOfTruth.of(
            reader = { userId ->
                userLocalDataSource.getUserFlow(userId)
                    .map { it?.let { DataSourceResult.Success(it) } } as Flow<DataSourceResult<User>?>
            },
            writer = { _: String, input: DataSourceResult<User> ->
                input.doOnSuccess {
                    userLocalDataSource.saveUser(it)
                }
            }
        )
    ).build()

    suspend fun getUser(id: String): DataSourceResult<User> = userStore.get(key = id)

    // In ViewModel
    viewModelScope.launch {
            try {
                val userResult = userRepo.getUser(userId)
                // This call is indefinitely suspended and no error is returned because no non-Loading and non-NoNewData result is returned
            } catch (e: Exception) {
                // Nothing is caught here. No exception is thrown
            }
    }
```

**Smartphone (please complete the following information):**
 - Device: Pixel 4XL
 - OS: Android 11
 - Store Version: com.dropbox.mobile.store:store4:4.0.2-KT15

**Comment by digitalbuddha** [comment]
Hmm sounds like a bug...From what I recall errors from fetcher propagate to the collector. Is it possible for you to make a failing test? There are some other tests that deal with network errors (on phone now can't find)

**Comment by TylerMcCraw** [comment]
Yep. I'll try to put together something tonight that can be used to test this.
But, also, if you're on paternity leave, don't spend your time on me! Go sleep or relax or something else.
(also, congrats on the baby!)

**Comment by digitalbuddha** [comment]
Thanks it's a blast. @alexio is going to take a look once you post a test.

**Comment by TylerMcCraw** [comment]
My description of the bug is completely off here.
But, also after having written a couple test cases based on my problem and what I expect, the tests are passing.
I've found that my reader and how it works with Realm has a problem in that it would never emit a null value if the object didn't exist in the cache.
This seems to be the cause of the issue I'm seeing in that the suspended function is never completed and so it's suspended indefinitely.
The `store.get(key)` function relies on data being returned. If nothing is ever returned, the collector will be in limbo forever.
I think that that is fine and I don't think it's Store's job to determine how to handle a reader never returning anything.

Closing the bug.
